src/clusterfuzz/_internal/system/archive.py

```diff
@@ -34,111 +36,316 @@
+  real_file_path = os.path.realpath(absolute_file_path)
 
-  def __init__(self, name, size, handle):
-    self.name = name
-    self.size = size
-    self.handle = handle
+  if real_file_path == output_dir:
+    # Workaround for https://bugs.python.org/issue28488.
+    # Ignore directories named '.'.
+    return False
 
+  if real_file_path != absolute_file_path:
+    logs.log_error('Directory traversal attempted while unpacking archive %s '
+                   '(file path=%s, actual file path=%s). Aborting.' %
+                   (archive_name, absolute_file_path, real_file_path))
+    return True
+  return False
 
-def iterator(archive_path,
-             archive_obj=None,
-             file_match_callback=None,
-             should_extract=True):
-  """Return an iterator for files in an archive. Extracts files if
-  |should_extract| is True."""
-  archive_type = get_archive_type(archive_path)
 
-  if not file_match_callback:
+@dataclasses.dataclass
+class ArchiveMemberInfo:
+  """Represents an archive member.
+  """
+  filename: str
+  is_dir: bool
+  file_size_bytes: int
+  mode: int
+
 
-    def file_match_callback_fallback(_):
-      return True
+class ArchiveReader(abc.ABC):
+  """Abstract class for representing an archive reader. Abstract methods must
+  be overriden.
+  """
 
-    file_match_callback = file_match_callback_fallback
+  @abc.abstractmethod
+  def list_files(self) -> typing.List[ArchiveMemberInfo]:
+    """Lists all members contained in the archives.
 
-  def maybe_extract(extract_func, info):
-    """Returns an extracted file or None if it is not supposed to be extracted.
+    Returns:
+        typing.List[ArchiveMemberInfo]: All the archive members
     """
-    if should_extract:
-      return extract_func(info)
-    return None
+    raise NotImplementedError
 
-  if archive_type == ArchiveType.ZIP:
-    try:
-      with zipfile.ZipFile(archive_obj or archive_path) as zip_file:
-        for info in zip_file.infolist():
-          if not file_match_callback(info.filename):
-            continue
+  @abc.abstractmethod
+  def extract(self, member, path, trusted=False) -> str:
+    """Extracts `member` out of the archive to the provided path.
+
+    Args:
+        member (str): the member name
+        path (str): the path where the member should be extracted.
+        Defaults to None.
+        trusted (bool, optional): whether the archive is trusted. Defaults to
+        False.
+
+    Returns:
+        str: the path to the extracted member
+    """
+    raise NotImplementedError
+
+  @abc.abstractmethod
+  def open(self, member):
+    """Opens `member`.
+
+    Args:
+        member (str): the member name
+
+    Returns:
+        [IO]: a file-like object that can be `read`.
+    """
+    raise NotImplementedError
+
+  @abc.abstractmethod
+  def close(self) -> None:
+    """Closes the archive.
+    """
+    raise NotImplementedError
+
+  @abc.abstractmethod
+  def extractall(self, path, members=None, trusted=False) -> None:
```

Comment:
Maybe this is modelled after the `zip` module API in which case it can stay, but for consistency it would be nice to keep this `snake_case` and name it `extract_all()`.